### PR TITLE
Add inference speed events

### PR DESCRIPTION
**Pull request by ambitious-octopus**
Still debugging this new feature.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Inference event collection now records the model’s parameter count alongside existing model metadata to provide additional context for speed analysis.

### 📊 Key Changes

- Imports `torch.nn` and `get_num_params` for parameter-count detection.
- Extracts the backend module from `run.model.model`.
- Adds an `nparams` field to event parameters when the backend is a PyTorch `nn.Module`.
- Sets `nparams` to `None` for non-PyTorch or unavailable backend modules.

### 🎯 Purpose & Impact

- Inference events now include model capacity metadata that can be used when analyzing inference speed.
- The change affects event data collection only; it does not modify the inference execution path.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ambitious-octopus, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an engineer will assist you soon. Please review the checklist below for a smooth integration of your changes, especially since this PR updates prediction event handling and documentation. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/callbacks/hub.py [hunk 1]
```diff
@@ -85,8 +85,35 @@ def on_val_start(validator):
...
+    except Exception:
+        model_params = None
+
+    # Per-image speed from the last processed batch
+    speed = None
+    results = getattr(predictor, "results", None)
+    if results:
+        speed = getattr(results[0], "speed", None)
+
+    events(predictor.args, predictor.device, backend=backend, imgsz=imgsz, model_params=model_params, speed=speed)
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/callbacks/hub.py]: ⚠️ **HIGH**: This callback queues a `predict` event for every batch. For multi-batch inference that will emit multiple telemetry records for one prediction command and can keep growing the in-memory queue until the rate limit trips. If the intent is one event per run, capture/update the latest metadata here and send a single event from `on_predict_end()` instead.

File: ultralytics/utils/events.py [hunk 2]
```diff
@@ -102,15 +116,28 @@ def __call__(self, cfg, device=None, backend=None) -> None:
...
+    def flush(self, timeout: float = 5.0) -> None:
+        """Block until the in-flight background send thread completes or times out.
+
+        Call this at the end of short-lived processes (e.g. single-image predict) so the
+        daemon thread is not killed before the POST request finishes.
+
+        Args:
+            timeout (float): Maximum seconds to wait for the thread to finish.
+        """
+        if self._thread is not None and self._thread.is_alive():
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/events.py:150]: ⚠️ **HIGH**: `flush()` only waits for an already-started background thread. If a predict run finishes before `rate_limit` elapses, `__call__()` leaves the event in `self.events` and never starts a thread, so `on_predict_end()` will return without sending anything. Short-lived inference runs would therefore lose their telemetry entirely unless `flush()` also posts any pending queue synchronously.

File: ultralytics/utils/callbacks/hub.py [hunk 3]
```diff
@@ -85,8 +85,35 @@ def on_val_start(validator):
...
+    """Run events on predict batch end, collecting imgsz, model params, and inference speed."""
+    model = getattr(predictor, "model", None)
+    backend = getattr(model, "backend", None)
+
+    # Image size from args
+    imgsz = getattr(predictor.args, "imgsz", None)
+
+    # Model parameter count (PyTorch models only)
+    try:
+        model_params = sum(p.numel() for p in model.parameters()) if model and hasattr(model, "parameters") else None
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/callbacks/hub.py:101]: 💡 **MEDIUM**: `sum(p.numel() for p in model.parameters())` runs on every batch, which adds O(parameter_count) work directly into the inference path. On larger models this becomes measurable overhead and can skew the very speed metric being reported. Cache this once per predictor/model instead of recomputing it for each batch.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Nice direction overall, especially deferring predict telemetry until result metadata is available and adding shutdown handling. The main issues are in the event lifecycle: short-lived predict runs currently drop queued events, batch-end reporting emits duplicate predict events for multi-batch runs, parameter counting adds avoidable per-batch overhead, and `debug_mode` is always reported as enabled.

💬 Posted 4 inline comments

File: ultralytics/utils/events.py [hunk 4]
```diff
@@ -60,6 +61,7 @@ def __init__(self) -> None:
             "env": ENVIRONMENT,
             "session_id": round(random.random() * 1e15),
             "engagement_time_msec": 1000,
+            "debug_mode":True,
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/events.py]: 💡 **MEDIUM**: `debug_mode` is hardcoded to `True`, so every event from normal usage will be tagged as a debug session. If this field is meant to reflect real runtime state, it should be derived from configuration or omitted when unavailable.
